x/ibc/03-connection/keeper/keeper_test.go

```diff
@@ -0,0 +1,169 @@
+	suite.Suite
+	apps map[string]App
+}
+
+func (suite *KeeperTestSuite) SetupTest() {
+	suite.apps = map[string]App{
+		ChainIDGaia1: NewApp(ChainIDGaia1),
+		ChainIDGaia2: NewApp(ChainIDGaia2),
+	}
+}
+
+func TestKeeperTestSuite(t *testing.T) {
+	suite.Run(t, new(KeeperTestSuite))
+}
+
+func (suite *KeeperTestSuite) TestSetAndGetConnection() {
+	gaia := suite.apps[ChainIDGaia1]
+
+	_, existed := gaia.connKeeper.GetConnection(gaia.ctx, ConnectionToGaia2)
+	suite.False(existed)
+
+	counterparty := types.NewCounterparty(ClientToGaia2, ConnectionToGaia1, gaia.connKeeper.GetCommitmentPrefix())
+	expConn := types.ConnectionEnd{
+		State:        types.INIT,
+		ClientID:     ClientToGaia1,
+		Counterparty: counterparty,
+		Versions:     types.GetCompatibleVersions(),
+	}
+	gaia.connKeeper.SetConnection(gaia.ctx, ConnectionToGaia2, expConn)
+	conn, existed := gaia.connKeeper.GetConnection(gaia.ctx, ConnectionToGaia2)
+	suite.True(existed)
+	suite.EqualValues(expConn, conn)
+}
+
+func (suite *KeeperTestSuite) TestSetAndGetClientConnectionPaths() {
+	gaia := suite.apps[ChainIDGaia1]
+
+	_, existed := gaia.connKeeper.GetClientConnectionPaths(gaia.ctx, ClientToGaia2)
+	suite.False(existed)
+
+	gaia.connKeeper.SetClientConnectionPaths(gaia.ctx, ClientToGaia2, types.GetCompatibleVersions())
+	paths, existed := gaia.connKeeper.GetClientConnectionPaths(gaia.ctx, ClientToGaia2)
+	suite.True(existed)
+	suite.EqualValues(types.GetCompatibleVersions(), paths)
+}
+
+func (suite *KeeperTestSuite) TestAddAndRemoveConnectionToClient() {
+	gaia := suite.apps[ChainIDGaia1]
+
+	//add connection to client
+	err := gaia.connKeeper.addConnectionToClient(gaia.ctx, ClientToGaia2, ConnectionToGaia1)
+	suite.Nil(err)
+	expConns := []string{ConnectionToGaia1}
+	conns, existed := gaia.connKeeper.GetClientConnectionPaths(gaia.ctx, ClientToGaia2)
+	suite.True(existed)
+	suite.EqualValues(expConns, conns)
+
+	//add connection to client once again
+	newConn := ConnectionToGaia1 + "1"
+	err = gaia.connKeeper.addConnectionToClient(gaia.ctx, ClientToGaia2, newConn)
+	suite.Nil(err)
+	expConns = append(expConns, newConn)
+	conns, existed = gaia.connKeeper.GetClientConnectionPaths(gaia.ctx, ClientToGaia2)
+	suite.True(existed)
+	suite.EqualValues(expConns, conns)
+
+	//remove connection from client
+	expConns = expConns[0:1]
+	err = gaia.connKeeper.removeConnectionFromClient(gaia.ctx, ClientToGaia2, newConn)
+	suite.Nil(err)
+	conns, existed = gaia.connKeeper.GetClientConnectionPaths(gaia.ctx, ClientToGaia2)
+	suite.True(existed)
+	suite.EqualValues(expConns, conns)
+
+	//remove connection from client again
+	err = gaia.connKeeper.removeConnectionFromClient(gaia.ctx, ClientToGaia2, newConn)
+	suite.NotNil(err)
+
+	err = gaia.connKeeper.removeConnectionFromClient(gaia.ctx, ClientToGaia2, ConnectionToGaia1)
+	suite.Nil(err)
+	conns, existed = gaia.connKeeper.GetClientConnectionPaths(gaia.ctx, ClientToGaia2)
+	suite.True(existed)
+	suite.Nil(conns)
+}
+
+type App struct {
```

Comment:
we can refactor this to use `simapp` instead

Reply:
done